File: TP4/Q03/Q03.java
import java.time.*;
import java.util.*;
import java.io.*;
import java.text.SimpleDateFormat;

class Pokemon{
	
	private int id;
	private int generation;
	private String name;
	private String description;
	private ArrayList<String> types;
	private ArrayList<String> abilities;
	private double weight;
	private double height;
	private int captureRate;
	private boolean legendary;
	private Date captureDate;

	Pokemon(int id, int generation, String name, String description, ArrayList<String>  types, ArrayList<String>  abilities, double weight, double height, int captureRate, boolean legendary, Date captureDate){
		this.id = id;
		this.generation = generation;
		this.name = name;
		this.description = description;
		this.types = types;
		this.abilities = abilities;
		this.weight = weight;
		this.height = height;
		this.captureRate = captureRate;
		this.legendary = legendary;
		this.captureDate = captureDate;
	}

	Pokemon(){
		this.id = 0;
		this.generation = 0;
		this.name = null;
		this.description = null;
		this.types = null;
		this.abilities = null;
		this.weight = 0.0;
		this.height = 0.0;
		this.captureRate = 0;
		this.legendary = false;
		this.captureDate = new Date();
	}


	public Pokemon clon(){
		return new Pokemon(id, generation, name, description, types, abilities, weight, height, captureRate, legendary, captureDate);
	}

	public static Pokemon make(String s) throws Exception {
		Pokemon result = new Pokemon();
		result.read(s);
		return result;
	}

	public static Pokemon[] readFile(String path) throws Exception {
		Pokemon[] result = new Pokemon[801];
		File fil = new File(path);
		Scanner csv = new Scanner(fil);
		String line = csv.nextLine();
		int i = 0;
		while(csv.hasNext()){
			line = csv.nextLine();
			result[i] = make(line);
			i++;
		}
		csv.close();
		return result;
	}


	//_____Getters_____
	public int getId(){
		return id;
	}
	public int getGeneration(){
		return generation;
	}
	public String getName(){
		return name;
	}
	public String getDescription(){
		return description;
	}
	public ArrayList<String> getTypes(){
		return types;
	}
	public ArrayList<String> getAbilities(){
		return abilities;
	}
	public double getWeight(){
		return weight;
	}
	public double getHeight(){
		return height;
	}
	public int getCaptureRate(){
		return captureRate;
	}
	public boolean isLegendary(){
		return legendary;
	}
	public Date getCaptureDate(){
		return captureDate;
	}

	//_____Setters_____
	public void setId (int id){
		this.id = id;
	}
	public void setGeneration (int generation){
		this.generation = generation;
	}
	public void setName (String name){
		this.name = name;
	}
	public void setDescription (String description){
		this.description = description;
	}
	public void setTypes (ArrayList<String> types){
		this.types = types;
	}
	public void setAbilities (ArrayList<String> abilities){
		this.abilities = abilities;
	}
	public void setWeight (double weight){
		this.weight = weight;
	}
	public void setHeight (double height){
		this.height = height;
	}
	public void setCaptureRate (int captureRate){
		this.captureRate = captureRate;
	}
	public void setLegendary (boolean legendary){
		this.legendary = legendary;
	}
	public void setCaptureDate (Date captureDate){
		this.captureDate = captureDate;
	}

	//____Print and Read____
	public void print(){
		System.out.print("[#" + id + " -> " + name + ": " + description + " - [");

		int reps = types.size();
		for(int i=0; i<reps-1; i++){
			System.out.print("'" + types.get(i) + "', ");
		}
		System.out.print("'" + types.get(reps-1) + "'] - [");

		reps = abilities.size();
		for(int i=0; i<reps-1; i++){
			//System.out.print("'" + abilities.get(i) + "', ");
			System.out.print(abilities.get(i) + ", ");
		}
		System.out.print(abilities.get(reps-1) + "] - ");

		System.out.print(weight + "kg - " + height + "m - " + captureRate + "% - ");

		if(isLegendary()) System.out.print("true ");
		else System.out.print("false ");

		System.out.print("- " + generation + " gen] - ");

		SimpleDateFormat formater = new SimpleDateFormat("dd/MM/yyyy");
		System.out.println(formater.format(captureDate));
	}

	public void read(String s) throws Exception {
		String format = "";
		boolean control = true;
		for(int i=0; i<s.length(); i++){
			char x = s.charAt(i);
			if(x == '"') control = !control;
			else if(x == ',' && control) format += ';';
			else if(x!='[' && x!=']') format += x;
		}
		String[] aux = format.split(";");
		//System.out.println(format);

		setId(Integer.parseInt(aux[0]));
		setGeneration(Integer.parseInt(aux[1]));
		setName(aux[2]);
		setDescription(aux[3]);

		String type1 = aux[4];
		String type2 = aux[5];
		ArrayList<String> tipos = new ArrayList<String>();
		tipos.add(type1);
		if(!type2.equals("")) tipos.add(type2);
		setTypes(tipos);
	

		boolean hasComma = false;
		ArrayList<String> abis = new ArrayList<String>();
		for(int i=0; i<aux[6].length() && !hasComma; i++)
			if(aux[6].charAt(i)==',') hasComma = true;
		if(!hasComma)
			abis.add(aux[6]);
		else{
			String[] abs = aux[6].split(", ");
			for(int i=0; i<abs.length; i++) abis.add(abs[i]);
		}
		setAbilities(abis);

		if(!aux[7].equals("")) setWeight(Double.parseDouble(aux[7]));
		if(!aux[8].equals("")) setHeight(Double.parseDouble(aux[8]));
		if(!aux[9].equals("")) setCaptureRate(Integer.parseInt(aux[9]));
		
		if(aux[10].equals("1")) setLegendary(true);

		SimpleDateFormat formater = new SimpleDateFormat("dd/MM/yyyy");
		setCaptureDate(formater.parse(aux[11]));
	}

}


//_____________________________________________________________________________________________________________________________________________________
//=====================================================================================================================================================
//_____________________________________________________________________________________________________________________________________________________


class TP{

	private Instant start;
	private Instant end;
	private int comparisons;
	private int movements;

	TP(){
		this.start = Instant.now();
		this.comparisons = 0;
		this.movements = 0;
	}

	public void comp(){
		this.comparisons++;
	}

	public void comp(int x){
		this.comparisons += x;
	}

	public void move(){
		this.movements++;
	}

	public void move(int x){
		this.movements += x;
	}

	void end(){
		this.end = Instant.now();
	}

	double diff(){
		return Duration.between(this.start, this.end).getNano() / 1000000000.0;
	}

	void print(String fileName) throws Exception{
		PrintWriter write = new PrintWriter(new FileWriter(fileName));
		write.printf("Matrícula: 855947\t");
		write.printf("Tempo de execução: " + diff() + "\t");
		write.printf("Comparações: " + comparisons + "\t");
		write.printf("Movimentações: " + movements);
		write.close();
	}

	void printSearch(String fileName) throws Exception{
		PrintWriter write = new PrintWriter(new FileWriter(fileName));
		write.printf("Matrícula: 855947\t");
		write.printf("Tempo de execução: " + diff() + "\t");
		write.printf("Comparações: " + comparisons);
		write.close();
	}
	
}

//_____________________________________________________________________________________________________________________________________________________
//=====================================================================================================================================================
//_____________________________________________________________________________________________________________________________________________________


class Node{
	public Pokemon p;
	public Node left;
	public Node right;

	Node(Pokemon p){
		this.p = p;
		left = null;
		right = null;
	}

	public int getFat(){
		int fatR = right==null ? 0 : 1+right.getFat();
		int fatL = left==null ? 0 : 1+left.getFat();
		return fatR - fatL;
	}

}

class Tree{
	public Node root;

	Tree(){
		root = null;
	}

	public Node rotLeft(Node n){
		Node noDir = n.right;
		Node noDirEsq = noDir.left;
		noDir.left = n;
		n.right = noDirEsq;
		return noDir;
	}

	public Node rotRight(Node n){
		Node noEsq = n.left;
		Node noEsqDir = noEsq.right;
		noEsq.right = n;
		n.left = noEsqDir;
		return noEsq;
	}

	public Node rotLR(Node n){
		n.left = rotLeft(n.left);
		return rotRight(n);
	}

	public Node rotRL(Node n){
		n.right = rotRight(n.left);
		return rotLeft(n);
	}

	public void printMid(){
		printMid(root);
	}

	public void balance(){
		balance(root);
	}

	public void balance(Node crr){
		int fac = crr.getFat();
		int son = 0;
		if(fac == 2){
			son = crr.right.getFat();
			crr = son==-1 ? rotRL(crr) : rotLeft(crr);
		}
		else if(fac == -2){
			son = crr.left.getFat();
			crr = son==1 ? rotLR(crr) : rotRight(crr);
		}
		if(crr.left != null) balance(crr.left);
		if(crr.right != null) balance(crr.right);
	}

	private void printMid(Node n){
		if(n.left != null) printMid(n.left);
		System.out.print(n.p.getName() + " ");
		if(n.right != null) printMid(n.right);
	}
	
	public void insert(Pokemon p){
		if(root == null) root = new Node(p);
		else if(p.getName().compareTo(root.p.getName()) < 0) insert(p, root.left, root);
		else if(p.getName().compareTo(root.p.getName()) > 0) insert(p, root.right, root);
		balance();
	}

	private void insert(Pokemon p, Node in, Node dad){
		if(in == null){
			if(p.getName().compareTo(dad.p.getName()) < 0) dad.left = new Node(p);
			else dad.right = new Node(p);
		}
		else if(p.getName().compareTo(in.p.getName()) < 0) insert(p, in.left, in);
		else insert(p, in.right, in);
	}

	public void search(String s, TP tp){
		System.out.println(s);
		System.out.print("raiz ");
		if(root == null) System.out.println("NAO");
		else{
			if(s.compareTo(root.p.getName()) == 0){
				System.out.println("SIM");
				tp.comp();
			}
			else if(s.compareTo(root.p.getName()) < 0){
				System.out.print("esq ");
				tp.comp();
				search(s, root.left, tp);
			}
			else{
				System.out.print("dir ");
				tp.comp();
				search(s, root.right, tp);
			}
		}
	}

	private void search(String s, Node n, TP tp){
		if(s.compareTo(n.p.getName()) == 0){
			System.out.println("SIM");
			tp.comp();
		}
		else if(s.compareTo(n.p.getName()) < 0){
			tp.comp();
			if(n.left == null) System.out.println("esq NAO");
			else{
				System.out.print("esq ");
				search(s, n.left, tp);
			}
		}
		else{
			if(n.right == null) System.out.println("dir NAO");
			else{
				System.out.print("dir ");
				search(s, n.right, tp);
			}
		}
	}

}


//_____________________________________________________________________________________________________________________________________________________
//=====================================================================================================================================================
//_____________________________________________________________________________________________________________________________________________________


public class Q03{

	static void swapPoke(Pokemon[] arr, int a, int b){
		Pokemon tmp = arr[a];
		arr[a] = arr[b];
		arr[b] = tmp;
	}

	public static void main(String[] args){
		try{

			Pokemon[] pokes = Pokemon.readFile("/tmp/pokemon.csv");
			Scanner sc = new Scanner(System.in);

			int x = 0;
			Tree t = new Tree();
			String input = sc.nextLine();
			while(!input.equals("FIM")){
				x = Integer.parseInt(input);
				t.insert(pokes[x-1]);
				input = sc.nextLine();
			}

			//t.printMid();


			/*
			TP tp = new TP();
			sort(using, tp);
			tp.end();
			for(int i=0; i<using.length; using[i++].print());
			tp.print("855947_mergesort.txt");
			*/

			TP tp = new TP();
			input = sc.nextLine();
			while(!input.equals("FIM")){
				t.search(input, tp);
				input = sc.nextLine();
			}
			tp.end();
			tp.printSearch("855947_arvoreBinaria.txt");

		} catch(FileNotFoundException e){
			System.out.println(e);
		} catch(Exception e){
			e.printStackTrace();
		}
	}
}

File: TP4/Q03/Q03.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>

#define strSize 80

void flag(){
	printf("FLAG\n");
}

int commaIndex(char* s){
	bool found = false;
	int i=0;
	for(; i<strlen(s) && !found; i++){
		if(s[i]=='.') found = true;
	}
	return i-1;
}

double parseDouble(char* s){
	int i = 0;
	double result = 0.0;
	int leng = strlen(s);
	int comma = commaIndex(s),
		amtLeft = leng-comma+2,
		j = comma-1;
	for(i=0; i<comma; i++){
		result += (double)((s[i]-48)*pow(10, j));
		j--;
	}
	j = 1;
	for(i=i+1; i<leng; i++){
		result += (double)((s[i]-48)/pow(10, j));
		j++;
	}
	return result;
}

int parseYear(char* s){
    int result = 0;
	int exp = 3;
	int i = 0;
    while(i<4){
        result += (int)(s[i]-48) * (int)pow(10.0, (double)exp);
        i++;
		exp--;
    }
    return result;
}

int parseInt(char* s){
    int result = 0;
    int i = strlen(s)-1,
        exp = 0;
    while(i>=0){
        result += (int)(s[i]-48) * (int)pow(10.0, (double)exp);
        i--;
		exp++;
    }
    return result;
}

typedef struct {
	int day;
	int month;
	int year;
} Date;

Date strToDate(char* s){
	char* d = strtok(s, "/");
	char* m = strtok(NULL, "/");
	char* y = strtok(NULL, "/");
	Date result;
	result.day = parseInt(d);
	result.month = parseInt(m);
	result.year = parseYear(y);
	return result;
}

char* dateToStr(Date i){
	char* result = malloc(40*sizeof(char));
	sprintf(result, "%02d/%02d/%04d", i.day, i.month, i.year);
	return result;
}

typedef struct {
	int n;
	char type1[strSize];
	char type2[strSize];
} Types;

typedef struct {
	int n;
	char ab1[strSize];
	char ab2[strSize];
	char ab3[strSize];
	char ab4[strSize];
	char ab5[strSize];
	char ab6[strSize];
} Abilities;

typedef struct {
	int id;
	int generation;
	char name[strSize];
	char description[strSize];
	Types types;
	Abilities abilities;
	double weight;
	double height;
	int captureRate;
	bool legendary;
	Date captureDate;
} Pokemon;

Pokemon* makeMon(char* line){
	Pokemon* result = calloc(1, sizeof(Pokemon));
	
	int leng = strlen(line);
	char* format = malloc(leng * sizeof(char));
	bool control = true;
	int j = 0;
	for(int i=0; i<leng; i++){
		if(line[i] == '"') control = !control;
		else if(line[i] == ',' && control) format[j++] = ';';
		else if(line[i] != '[' && line[i] != ']') format[j++] = line[i];
	}
	//printf("%s\n", format);

	char** aux = calloc(12, sizeof(char*));
	int nSplits=0;
	for(aux[nSplits++]=strtok(format,";"); aux[nSplits]=strtok(NULL,";"); nSplits++);

	//for(int x=0; x<nSplits; printf("%s\n", aux[x++]));

	result->id = parseInt(aux[0]);
	result->generation = parseInt(aux[1]);
	strcpy(result->name, aux[2]);
	strcpy(result->description, aux[3]);

	result->types.n = nSplits==11 ? 1 : 2;
	strcpy(result->types.type1, aux[4]);
	if(result->types.n == 2) strcpy(result->types.type2, aux[5]);

	bool commaInAbilities = false;
	int posAbs = nSplits==10 ? 6 : nSplits==11 ? 5 : 6;
	for(int j=0; j<strlen(aux[posAbs]); j++)
		if(aux[posAbs][j]==',') commaInAbilities = true;
	if(!commaInAbilities){
		result->abilities.n = 1;
		strcpy(result->abilities.ab1, aux[posAbs]);
	}
	else{

		//remove spaces after commas
		char* formt = malloc(strSize * sizeof(char));
		strcpy(formt, aux[posAbs]);
		for(int x=0; x<strlen(formt); x++)
			if(formt[x]==' ' && formt[x-1]==',') formt[x] = ',';

		char* buffer;
		buffer = strtok(formt, ",");
		result->abilities.n = 1;
		strcpy(result->abilities.ab1, buffer);

		if(buffer = strtok(NULL, ",")){
			result->abilities.n = 2;
			strcpy(result->abilities.ab2, buffer);

			if(buffer = strtok(NULL, ",")){
				result->abilities.n = 3;
				strcpy(result->abilities.ab3, buffer);

				if(buffer = strtok(NULL, ",")){
					result->abilities.n = 4;
					strcpy(result->abilities.ab4, buffer);

					if(buffer = strtok(NULL, ",")){
						result->abilities.n = 5;
						strcpy(result->abilities.ab5, buffer);

						if(buffer = strtok(NULL, ",")){
							result->abilities.n = 6;
							strcpy(result->abilities.ab6, buffer);
						}
					}
				}
			}
		}
	}


	if(nSplits>10){
		result->weight = parseDouble(aux[7-(12-nSplits)]);
		result->height = parseDouble(aux[8-(12-nSplits)]);
	}
	else{
		result->weight = 0.0;
		result->height = 0.0;
	}

	result->captureRate = parseInt(aux[9-(12-nSplits)]);

	if(strcmp(aux[10-(12-nSplits)], "1") == 0) result->legendary = true;
	else result->legendary = false;

	//printf("%s\n", aux[11-(12-nSplits)]);
	result->captureDate = strToDate(aux[11-(12-nSplits)]);


	return result;
}

void printMon(Pokemon* poke){
	printf("[#%d -> %s: %s - [", poke->id, poke->name, poke->description);

	if(poke->types.n == 1) printf("'%s'] - [", poke->types.type1);
	else printf("'%s', '%s'] - [", poke->types.type1, poke->types.type2);

	if(poke->abilities.n == 1)
		printf("%s] - ", poke->abilities.ab1);
	else if(poke->abilities.n == 2)
		printf("%s, %s] - ", poke->abilities.ab1, poke->abilities.ab2);
	else if(poke->abilities.n == 3)
		printf("%s, %s, %s] - ", poke->abilities.ab1, poke->abilities.ab2, poke->abilities.ab3);
	else if(poke->abilities.n == 4)
		printf("%s, %s, %s, %s] - ", poke->abilities.ab1, poke->abilities.ab2, poke->abilities.ab3, poke->abilities.ab4);
	else if(poke->abilities.n == 5)
		printf("%s, %s, %s, %s, %s] - ", poke->abilities.ab1, poke->abilities.ab2, poke->abilities.ab3, poke->abilities.ab4, poke->abilities.ab5);
	else if(poke->abilities.n == 6)
		printf("%s, %s, %s, %s, %s, %s] - ", poke->abilities.ab1, poke->abilities.ab2, poke->abilities.ab3, poke->abilities.ab4, poke->abilities.ab5, poke->abilities.ab6);

		printf("%.1lfkg - %.1lfm - %d%c - ", poke->weight, poke->height, poke->captureRate, '%');

		if(poke->legendary) printf("true ");
		else printf("false ");

		printf("- %d gen] - ", poke->generation);

		char* capDate = dateToStr(poke->captureDate);
		printf("%s\n", capDate);
}

Pokemon** readFile(const char* path){
	Pokemon** result = calloc(803, sizeof(Pokemon*));
	FILE* csv = fopen(path, "rt");
	char* trash = malloc(1000 * sizeof(char));
	fgets(trash, 999, csv);
	free(trash);
	for(int i=0; i<801; i++){
		char* buffer = malloc(1000 * sizeof(char));
		fgets(buffer, 999, csv);
		result[i] = makeMon(buffer);
		free(buffer);
	}
	fclose(csv);
	return result;
}

//_____________________________________________________________________________________________________________________________________________________
//=====================================================================================================================================================
//_____________________________________________________________________________________________________________________________________________________


double diff(clock_t start, clock_t end){
	return ((double)(end - start)) / CLOCKS_PER_SEC;
}

void logTP(const char* fileName, double time, int comps, int moves){
	FILE* fil = fopen(fileName, "w");
	if(fil){
		fprintf(fil, "Matrícula: 855947\t");
		fprintf(fil, "Tempo de Execução: %lf\t", time);
		fprintf(fil, "Comparações: %d\t", comps);
		fprintf(fil, "Movimentações: %d", moves);
		fclose(fil);
	}
	else printf("Erro ao abrir o arquivo");
}

void logSearch(const char* fileName, double time, int comps){
	FILE* fil = fopen(fileName, "w");
	if(fil){
		fprintf(fil, "Matrícula: 855947\t");
		fprintf(fil, "Tempo de Execução: %lf\t", time);
		fprintf(fil, "Comparações: %d", comps);
		fclose(fil);
	}
	else printf("Erro ao abrir o arquivo");
}

//_____________________________________________________________________________________________________________________________________________________
//=====================================================================================================================================================
//_____________________________________________________________________________________________________________________________________________________


typedef struct Node Node;
typedef struct Node{
	Pokemon* p;
	Node* left;
	Node* right;
	int fac;
} Node;

Node* newNode(Pokemon* x){
	Node* ret = calloc(1,sizeof(Node));
	ret->left = ret->right = NULL;
	ret->fac = 0;
	ret->p = x;
	return ret;
}

typedef struct Tree{
	Node* root;
} Tree;

Tree* newTree(){
	Tree* ret = calloc(1,sizeof(Tree));
	ret->root = NULL;
	return ret;
}

void rotL(Node** n){
	Node* noDir = (*n)->right;
	Node* noDirEsq = noDir->left;
	noDir->left = *n;
	(*n)->right = noDirEsq;
	*n = noDir;
}

void rotR(Node** n){
	Node* noEsq = (*n)->left;
	Node* noEsqDir = noEsq->right;
	noEsq->right = *n;
	(*n)->left = noEsqDir;
	*n = noEsq;
}

void rotLR(Node** n){
	rotL(&(*n)->left);
	rotR(n);
}

void rotRL(Node** n){
	rotR(&(*n)->right);
	rotL(n);
}

void insertRec(Pokemon* in, Node** i){
	if(*i == NULL) *i = newNode(in);
	else if(0 > strcmp(in->name, (*i)->p->name)){
		insertRec(in, &(*i)->left);
	}
	else{
		insertRec(in, &(*i)->right);
	}
}

void insert(Tree* t, Pokemon* in){
	insertRec(in, &t->root);
}

void searchRec(char* s, Node* n){
	if(0 == strcmp(s, n->p->name)){
		printf("SIM\n");
	}
	else if(0 > strcmp(s, n->p->name)){
		if(!n->left) printf("esq NAO\n");
		else{
			printf("esq ");
			searchRec(s, n->left);
		}
	}
	else{
		if(!n->right) printf("dir NAO\n");
		else{
			printf("dir ");
			searchRec(s, n->right);
		}
	}
}

void search(char* s, Node* n){
	printf("%s\n", s);
	printf("raiz ");
	if(n == NULL) printf("NAO\n");
	else{
		if(0==strcmp(s, n->p->name)){
			printf("SIM");
		}
		else if(0 > strcmp(s, n->p->name)){
			printf("esq ");
			searchRec(s, n->left);
		}
		else{
			printf("dir ");
			searchRec(s, n->right);
		}
	}
}



//_____________________________________________________________________________________________________________________________________________________
//=====================================================================================================================================================
//_____________________________________________________________________________________________________________________________________________________


int comps = 0;
int moves = 0;

bool equals(const char* s1, const char* s2){
	return 0==strcmp(s1, s2);
}

void swapInt(int* arr, int a, int b){
	int tmp = arr[a];
	arr[a] = arr[b];
	arr[b] = tmp;
}

int main(){
	Pokemon** pokes = readFile("/tmp/pokemon.csv");

	int i=0;
	int* usingIds = malloc(100*sizeof(int));
	
	char* input = malloc(20*sizeof(char));
	scanf(" %s", input);
	while(!equals(input, "FIM")){
		usingIds[i] = parseInt(input);
		i++;
		free(input);
		input = malloc(20*sizeof(char));
		scanf(" %s", input);
	}
	free(input);

	Tree* t = newTree();
	for(int j=0; j<i; j++){
		insert(t, pokes[usingIds[j]-1]);
	}
	free(usingIds);

	input = malloc(40*sizeof(char));
	scanf(" %s", input);
	while(!equals(input, "FIM")){

		search(input, t->root);

		free(input);
		input = malloc(40*sizeof(char));
		scanf(" %s", input);
	}
	free(input);

	/*
	clock_t start = clock();
	sort(using, i);
	for(int j=0; j<i; printMon(using[j++]));
	clock_t end = clock();
	*/

	free(pokes);
	//logTP("855947_bolha.txt", diff(start, end), comps, moves);

	return 0;
}
